contracts/modules/LoanClosings/LoanClosings.sol

```diff
@@ -6,88 +6,431 @@
         payable
         nonReentrant
         returns (
             uint256 loanCloseAmount,
-            uint256 seizedAmount,
-            address seizedToken
+            uint256 withdrawAmount,
+            address withdrawToken
         )
     {
-        return _liquidate(
+        return _closeWithDeposit(
             loanId,
             receiver,
-            closeAmount
+            depositAmount,
+            loanDataBytes
         );
     }
 
-    function closeWithDeposit(
+    function closeWithSwap(
         bytes32 loanId,
         address receiver,
-        uint256 depositAmount) // denominated in loanToken
+        uint256 swapAmount, // denominated in collateralToken
+        bool returnTokenIsCollateral, // true: withdraws collateralToken, false: withdraws loanToken
+        bytes memory loanDataBytes)
         public
-        payable
         nonReentrant
         returns (
             uint256 loanCloseAmount,
             uint256 withdrawAmount,
             address withdrawToken
         )
     {
-        return _closeWithDeposit(
+        return _closeWithSwap(
             loanId,
             receiver,
-            depositAmount
+            swapAmount,
+            returnTokenIsCollateral,
+            loanDataBytes
         );
     }
 
-    function closeWithSwap(
+    function _checkPermit(address token, bytes memory loanDataBytes) internal {
+        if(abi.decode(loanDataBytes, (uint128)) & WITH_PERMIT != 0) {
+            (uint128 f, bytes[] memory payload) = abi.decode(
+                loanDataBytes,
+                (uint128, bytes[])
+            );
+            (address owner, address spender, uint value, uint deadline, uint8 v, bytes32 r, bytes32 s) = abi.decode(payload[2], (address, address, uint, uint, uint8, bytes32, bytes32));
+            IERC20Permit(token).permit(owner, spender, value, deadline, v, r, s);
+        }
+    }
+
+    function _closeWithDeposit(
         bytes32 loanId,
         address receiver,
-        uint256 swapAmount, // denominated in collateralToken
-        bool returnTokenIsCollateral, // true: withdraws collateralToken, false: withdraws loanToken
+        uint256 depositAmount, // denominated in loanToken
         bytes memory loanDataBytes)
-        public
-        nonReentrant
+        internal
+        pausable
         returns (
             uint256 loanCloseAmount,
             uint256 withdrawAmount,
             address withdrawToken
         )
     {
-        return _closeWithSwap(
+        require(depositAmount != 0, "depositAmount == 0");
+
+        Loan memory loanLocal = loans[loanId];
+        _checkAuthorized(
+            loanLocal.id,
+            loanLocal.active,
+            loanLocal.borrower
+        );
+
+        if (receiver == address(0)) {
+            receiver = msg.sender;
+        }
+
+        LoanParams memory loanParamsLocal = loanParams[loanLocal.loanParamsId];
+
+        _checkPermit(loanParamsLocal.collateralToken, loanDataBytes);
```

Comment:
should check permit on loanToken, not collateralToken.

Reply:
good point

Reply:
fixed